Yank/repex.py

```diff
@@ -2909,10 +3232,29 @@ def _compute_replica_energies(self, replica_id):
         # Update potential energy by temperature.
         for thermodynamic_state_id, thermodynamic_state in enumerate(self._thermodynamic_states):
             beta = 1.0 / (mmtools.constants.kB * thermodynamic_state.temperature)
-            replica_energies[replica_id, thermodynamic_state_id] = beta * reference_reduced_potential
+            energy_thermodynamic_states[thermodynamic_state_id] = beta * reference_reduced_potential
+
+        # Since no assumptions can be made about the unsampled thermodynamic states, do it the hard way
+        for unsampled_id, state in enumerate(self._unsampled_states):
+            if unsampled_id == 0 or not state.is_state_compatible(context_state):
```

Comment:
It would be good to refactor the duplicated code into a `MultiStateSampler` private functions.